File: src/contest/w428/CountBeautifulSplitsInAnArray.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package contest.w428;

import java.util.HashSet;
import java.util.Set;

/**
 * LeetCode 3388
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/12/15 11:18
 */
public interface CountBeautifulSplitsInAnArray {
    int beautifulSplits(int[] nums);

    class Solution1 implements CountBeautifulSplitsInAnArray {
        @Override
        public int beautifulSplits(int[] nums) {
            if (nums.length < 3) {
                return 0;
            }

            Set<String> set = new HashSet<>();
            for (int left = 0; left < nums.length - 1; left++) {
                int right = left + 1;
                while (right < nums.length) {
                    if (isPrefix(nums, left, right)) {
                        if (left == 0) {
                            int midRight = right * 2 - left;
                            while (midRight < nums.length) {
                                set.add(right + "-" + midRight);
                                midRight++;
                            }
                        } else {
                            set.add(left + "-" + right);
                        }
                    }
                    right++;
                }
            }

            return set.size();
        }

        private boolean isPrefix(int[] nums, int left, int right) {
            if (left < 0 || right > nums.length - 1 || left >= right) {
                return false;
            }
            int i = left;
            int j = right;
            while (j < nums.length && i < right) {
                if (nums[i++] != nums[j++]) {
                    return false;
                }
            }
            return i == right;
        }
    }

    class Solution2 implements CountBeautifulSplitsInAnArray {
        @Override
        public int beautifulSplits(int[] nums) {
            int n = nums.length;
            if (n < 3) {
                return 0; // We need at least 3 elements to make a valid split
            }

            long mod = 1_000_000_007;
            long base = 31;

            // Compute prefix hashes
            long[] prefixHash = new long[n + 1]; // Hash for prefix [0..i)
            long[] pow = new long[n + 1]; // Store powers of base

            pow[0] = 1;
            for (int i = 0; i < n; i++) {
                pow[i + 1] = (pow[i] * base) % mod;
                prefixHash[i + 1] = (prefixHash[i] * base + nums[i]) % mod;
            }

            int count = 0;

            // Iterate over possible i and j such that 1 <= i < j < n
            for (int i = 1; i < n - 1; i++) {
                for (int j = i + 1; j < n; j++) {
                    boolean valid = false;

                    // Check if nums1 is a prefix of nums2
                    if (isPrefix(prefixHash, 0, i, i, j, mod, pow)) {
                        valid = true;
                    }

                    // Check if nums2 is a prefix of nums3
                    if (!valid && isPrefix(prefixHash, i, j, j, n, mod, pow)) {
                        valid = true;
                    }

                    if (valid) {
                        count++;
                    }
                }
            }

            return count;
        }

        // Helper function to compare hashes of two subarrays
        private boolean isPrefix(long[] hash, int start1, int end1, int start2, int end2, long mod, long[] pow) {
            int len1 = end1 - start1;
            int len2 = end2 - start2;

            if (len1 > len2) {
                return false;
            }

            long hash1 = (hash[end1] - (hash[start1] * pow[len1]) % mod + mod) % mod;
            long hash2 = (hash[start2 + len1] - (hash[start2] * pow[len1]) % mod + mod) % mod;

            return hash1 == hash2;
        }
    }
}

File: src/contest/w428/CountBeautifulSplitsInAnArray.go
package w428

const mod = int(1e9 + 7)
const base = 31

// LeetCocde 3388
func beautifulSplits(nums []int) int {
	n := len(nums)
	if n < 3 {
		return 0
	}

	pow := make([]int, n+1)
	prefixHash := make([]int, n+1)

	pow[0] = 1
	for i := 1; i < n+1; i++ {
		pow[i] = pow[i-1] * base % mod
		prefixHash[i] = (prefixHash[i-1]*base + nums[i-1]) % mod
	}

	count := 0
	for i := 1; i < n-1; i++ {
		for j := i + 1; j < n; j++ {
			isPrefix := false

			if isPrefixSum(0, i, i, j, prefixHash, pow) {
				isPrefix = true
			}

			if isPrefixSum(i, j, j, n, prefixHash, pow) {
				isPrefix = true
			}

			if isPrefix {
				count++
			}
		}
	}
	return count
}

func isPrefixSum(start1, end1, start2, end2 int, hash, pow []int) bool {
	len1 := end1 - start1
	len2 := end2 - start2
	if len1 > len2 {
		return false
	}

	hash1 := (hash[end1] - hash[start1]*pow[len1]%mod + mod) % mod
	hash2 := (hash[start2+len1] - hash[start2]*pow[len1]%mod + mod) % mod
	return hash1 == hash2
}
